ActiveSagaInstance enhancements

boblangley opened the pull request:
Connects to https://github.com/Particular/PlatformDevelopment/issues/677

Moving SagaAudit to utilize strongly typed stages for it's behaviors [instead of `InsertBefore` and `InsertAfter`](https://github.com/Particular/NServiceBus/issues/3582) the handler is no longer exposed to one of the behaviors. Removing `obsolete` allows us to access the saga `Instance` in the plugin.

Likewise the `InsertBefore`/`InsertAfter` were used to try to measure the Saga lifespan as part of the auditing process. Adding the `DateTime` properties `Created` and `Modified` along with `MarkAsModified` give us those metrics without the need to specify specific ordering in the pipeline and are more accurate.

boblangley commented:
I left the 2 commits because the 2 properties may be controversial. I will squash if we agree to keep them.

Adding them makes the SagaAudit metrics much simpler to implement and may provide a basis for future PerformanceCounters distinct from the message processing time since this covers end-to-end saga time including persistence.

boblangley commented:
I am confident, rebased and squashed.

boblangley commented:
@Particular/nservicebus-maintainers Please review my shenanigans.

danielmarbach commented on src/NServiceBus.Core/Sagas/ActiveSagaInstance.cs:
```diff
@@ -83,6 +96,7 @@ internal void AttachExistingEntity(IContainSagaData loadedEntity)
         void AttachEntity(IContainSagaData sagaEntity)
         {
             sagaId = sagaEntity.Id;
+            Modified = currentUtcDateTimeProvider();
```
Use `MarkAsModified`?

danielmarbach commented on src/NServiceBus.Core/Sagas/ActiveSagaInstance.cs:
```diff
@@ -110,8 +124,13 @@ void AttachEntity(IContainSagaData sagaEntity)
         internal void MarkAsNotFound()
         {
             NotFound = true;
+            Modified = currentUtcDateTimeProvider();
```
Use `MarkAsModified`?

danielmarbach commented:
For me good but I'm biased. So I want others to have a look at it

timbussmann commented:
I don't really like it since the code calls `MarkAsModified()` from inside as well as outside the class, it seems a bit leaky to me.

Won't block this though, since I can't come up with a better (easy) solution. The best way would probably be to hook into the new notifications mechanism and raise events for saga updates & completions, this would actually be even interesting for other subscribers.

danielmarbach commented:
@timbussmann The whole `ActiveSagaInstance` is a weird concept since it is not really tight into the lifecycle managed by the saga persistence behavior and the behavior itself. So the whole class is leaky. This doesn't make it worse but enables us to get the plugin going. The notifications way would be much better but also explode the scope since we would need to make that concept public.

timbussmann commented:
> The notifications way would be much better but also explode the scope since we would need to make that concept public.

I know, that's why I'm (somewhat) ok with the changes ;)

> This doesn't make it worse

I disagree, but again I don't have a better solution at hand.

My main concern is that everything we're introducing here can only be removed again in the next major version and we're doing this only because we dislike InsertBefore/After which is still around and also can only be removed on a major version. So we're abandoning code we want to get rid off by adding code we want to get rid off.

But feel free to merge.

timbussmann commented on src/NServiceBus.Core/Sagas/ActiveSagaInstance.cs:
```diff
@@ -110,8 +124,13 @@ void AttachEntity(IContainSagaData sagaEntity)
+            MarkAsModified();
         }
 
+        internal void MarkAsModified()
```
it's kinda weird that `MarkAsModified` is internal while `Modified` has a public setter?

andreasohlund replied:
Yea, can we make the setter private?

andreasohlund commented:
You do have a point @timbussmann but since `ActiveSagaInstance` is public anyway I'd say adding a few members is a small price to pay to avoid using the smelly before/after?

When we move to notifications/whatever we're most likely going to obsolete the whole class anyway so I think this is a price we can pay?

timbussmann commented:
> When we move to notifications/whatever we're most likely going to obsolete the whole class anyway

It didn't appear to me that the rest on that class is somehow related to notifications, but :+1: to removing it from the public API.

andreasohlund commented:
>It didn't appear to me that the rest on that class is somehow related to
notifications, but [image: :+1:] to removing it from the public API.

Yes, but that would most likely happen in vNext

On Thu, Mar 17, 2016 at 2:12 PM, Tim Bussmann <notifications@github.com>
wrote:

> When we move to notifications/whatever we're most likely going to obsolete
> the whole class anyway
>
> It didn't appear to me that the rest on that class is somehow related to
> notifications, but [image: :+1:] to removing it from the public API.
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/Particular/NServiceBus/pull/3607#issuecomment-197873052>
>

boblangley commented:
Setters are now `private` (or removed). I agree with @timbussmann re: Internal + External use of method, so I reverted. I am happy either way, but I will be opinionated.

@timbussmann Ready to merge?

timbussmann commented:
@boblangley I think there are two comments from daniel left.

boblangley commented:
Moved setting `Modified` using the `Func<DateTime>` to private method so it is DRY. I believe this resolve the concerns raised by @danielmarbach

I split, as originally proposed by @danielmarbach, `MarkAsModified()` into `Completed()` and `Updated()` and their calls. This plugs the leaks and should resolve the concern raised by @timbussmann
